router/throttler/throttler.go

```diff
@@ -1,173 +1,89 @@
-	var destLevelLimitReached bool
-	if throttler.destLimiter.enabled {
-		destKey := throttler.getDestKey(destID)
-		limitStatus, err := throttler.destLimiter.ratelimiter.Check(destKey, currentTime)
-		if err != nil {
-			// TODO: handle this
-			pkgLogger.Errorf(`[[ %s-router-throttler: Error checking limitStatus: %v]]`, throttler.destinationName, err)
-		} else {
-			destLevelLimitReached = limitStatus.IsLimited
-		}
-	}
-
-	var userLevelLimitReached bool
-	if !destLevelLimitReached && throttler.userLimiter.enabled {
-		userKey := throttler.getUserKey(destID, userID)
-		limitStatus, err := throttler.userLimiter.ratelimiter.Check(userKey, currentTime)
-		if err != nil {
-			// TODO: handle this
-			pkgLogger.Errorf(`[[ %s-router-throttler: Error checking limitStatus: %v]]`, throttler.destinationName, err)
-		} else {
-			userLevelLimitReached = limitStatus.IsLimited
-		}
-	}
-
-	return destLevelLimitReached || userLevelLimitReached
+type throttlingConfig struct {
+	enabled bool
+	limit   int64
+	window  time.Duration
 }
 
-// Inc increases the destLimiter and userLimiter counters.
-// If destID or userID passed is empty, we don't increment the counters.
-func (throttler *HandleT) Inc(destID, userID string, currentTime time.Time) {
-	if throttler.destLimiter.enabled && destID != "" {
-		destKey := throttler.getDestKey(destID)
-		_ = throttler.destLimiter.ratelimiter.Inc(destKey, currentTime)
+func (c *throttlingConfig) readThrottlingConfig(destName, destID string) {
+	if config.IsSet(fmt.Sprintf(`Router.throttler.%s.%s.limit`, destName, destID)) {
+		config.RegisterInt64ConfigVariable(
+			0, &c.limit, false, 1, fmt.Sprintf(`Router.throttler.%s.%s.limit`, destName, destID),
+		)
+	} else {
+		config.RegisterInt64ConfigVariable(0, &c.limit, false, 1, fmt.Sprintf(`Router.throttler.%s.limit`, destName))
 	}
-	if throttler.userLimiter.enabled && userID != "" {
-		userKey := throttler.getUserKey(destID, userID)
-		_ = throttler.userLimiter.ratelimiter.Inc(userKey, currentTime)
-	}
-}
 
-// Dec decrements the destLimiter and userLimiter counters by count passed
-// If destID or userID passed is empty, we don't decrement the counters.
-func (throttler *HandleT) Dec(destID, userID string, count int64, currentTime time.Time, atLevel string) {
-	if throttler.destLimiter.enabled && destID != "" && (atLevel == ALL_LEVELS || atLevel == DESTINATION_LEVEL) {
-		destKey := throttler.getDestKey(destID)
-		_ = throttler.destLimiter.ratelimiter.Dec(destKey, count, currentTime)
+	if config.IsSet(fmt.Sprintf(`Router.throttler.%s.%s.timeWindow`, destName, destID)) ||
+		config.IsSet(fmt.Sprintf(`Router.throttler.%s.%s.timeWindowInS`, destName, destID)) {
+		config.RegisterDurationConfigVariable(
+			0, &c.window, false, time.Second,
+			fmt.Sprintf(`Router.throttler.%s.%s.timeWindow`, destName, destID),
+			fmt.Sprintf(`Router.throttler.%s.%s.timeWindowInS`, destName, destID),
```

Comment:
All calls to `config.RegisterXXX` keep pointers to resources which we might desire to release at some point but it won't be possible, i.e. are potential sources of memory leaks.
Let's avoid them wherever possible
```suggestion
		c.window = config.GetDuration(fmt.Sprintf(`Router.throttler.%s.%s.timeWindow`, destName, destID), 0, time.Second)
```